awscli/customizations/cloudformation/deploy.py

```diff
@@ -359,6 +356,79 @@ def merge_parameters(self, template_dict, parameter_overrides):
 
         return parameter_values
 
+    def _validate_cf_params(self, param):
+        if 'ParameterKey' in param and 'ParameterValue' in param:
+            if len(param.keys()) > 2:
+                raise ParamValidationError(
+                    'CloudFormation like parameter JSON should have format '
+                    '[{"ParameterKey": "string", "ParameterValue": "string"}]')
+        return True
+
+    def _cf_param_parser(self, data):
+        # Parse parameter_overrides if they were given in
+        # CloudFormation params file format
+        # [{
+        #   "ParameterKey": "string",
+        #   "ParameterValue": "string",
+        # }]
+        try:
+            return {
+                param['ParameterKey']: param['ParameterValue']
+                for param in data if self._validate_cf_params(param)
+            }
+        except (KeyError, TypeError):
+            return None
+
+    def _codepipeline_param_parser(self, data):
+        # Parse parameter_overrides if they were given in
+        # CodePipeline params file format
+        # {
+        #     "Parameters": {
+        #         "ParameterKey": "ParameterValue"
+        #     }
+        # }
+        if isinstance(data, dict):
+            return data.get('Parameters', None)
+
+    def _parse_input_as_json(self, arg_value):
+        # In case of inline json input it'll be list where json string
+        # will be the first element
+        if arg_value:
+            if isinstance(arg_value, str):
+                return json.loads(arg_value)
+            try:
+                return json.loads(arg_value[0])
+            except json.JSONDecodeError:
+                return None
+
+    def parse_parameter_overrides(self, arg_value):
+        data = self._parse_input_as_json(arg_value)
+        if data is not None:
+            parsers = [
+                self._cf_param_parser,
+                self._codepipeline_param_parser,
+
+            ]
+            result = functools.reduce(
+                lambda a, parser: a or parser(data),
+                parsers, None
+            )
+            # In case it was in deploy command format
+            # but was in file
+            if result is None:
+                return self.parse_key_value_arg(
+                    data,
+                    self.PARAMETER_OVERRIDE_CMD
+                )
+            return result
```

Comment:
So I'm not entire sure we should be relying on the old parsing behavior as the final attempt and not provide any helpful error message after it. Specifically playing around with this I got some really weird looking errors.

In this example, I accidentally misspelled `ParameterValue` as `ParameterVlue`
```
$ aws cloudformation deploy --stack-name test-cfn --template-file template.json --parameter-overrides "[{\"ParameterKey\":\"BucketName\",\"ParameterVlue\":\"kyleknap-test-cfn-bucket\"},{\"ParameterKey\":\"Tag1value\",\"ParameterValue\":\"awesomeBucket\"}]"

'dict' object has no attribute 'split'

``` 
Then even in the test case for just a generally bad structure, this message is confusing:
```
$ aws cloudformation deploy --stack-name test-cfn --template-file template.json --parameter-overrides '{"SomeKey":[{"RedundantKey": "RedundantValue"}]}'

['SomeKey'] value passed to --parameter-overrides must be of format Key=Value
```
In general, it might make sense to abstract this logic a bit into classes/this format:
```python
class BaseParameterOverrideParser:
     def can_parse(self, data):
           # Returns true/false if it can parse

    def parse(self, data):
         # Return the properly formatted parameter dictionary

# Example parser
class CodePipelineParameterOverrideParser(BaseParameterOverrideParser):
     def can_parse(self, data):
         return isinstance(data, dict) and 'Parameters' in data

     def parse(self, data):
         return data['Parameters']


# In the DeployCommand class
def parse_parameter_overrides(self, arg_value):
     # After checking if it is json:
     parsers = [
          CfnParameterOverrideParser(),  # maybe find better names here
          CodePipelineParameterOverrideParser(),
          StringEqualsParameterOverrideParser()  
     ]
     for parser in parsers:
         if parser.can_parse(data):
            return parser.parse(data)
     raise InvalidParameterOverrideFormat('Show helpful information about the accepted formats')
```
The advantages with this approach is that:

* We can avoid weird/unexpected errors from trying to parse a format that will not work.
* We can throw a more helpful error about generally incorrect parameter formats instead of relying on any one specific parser errors.